lib/src/source/hosted.dart

```diff
@@ -152,55 +153,115 @@ class HostedSource extends Source {
+    final desc = _asDescription(description);
+    return _serializedDescriptionFor(desc.packageName, desc.uri);
+  }
+
   @override
   String formatDescription(description) =>
-      'on ${_parseDescription(description).last}';
+      'on ${_asDescription(description).uri}';
 
   @override
   bool descriptionsEqual(description1, description2) =>
-      _parseDescription(description1) == _parseDescription(description2);
+      _asDescription(description1) == _asDescription(description2);
 
   @override
-  int hashDescription(description) => _parseDescription(description).hashCode;
+  int hashDescription(description) => _asDescription(description).hashCode;
 
   /// Ensures that [description] is a valid hosted package description.
   ///
-  /// There are two valid formats. A plain string refers to a package with the
-  /// given name from the default host, while a map with keys "name" and "url"
-  /// refers to a package with the given name from the host at the given URL.
+  /// Simple hosted dependencies only consist of a plain string, which is
+  /// resolved against the default host. In this case, [description] will be
+  /// null.
+  ///
+  /// Hosted dependencies may also specify a custom host from which the package
+  /// is fetched. There are two syntactic forms of those dependencies:
+  ///
+  ///  1. With an url and an optional name in a map: `hosted: {url: <url>}`
+  ///  2. With a direct url: `hosted: <url>`
   @override
-  PackageRef parseRef(String name, description, {String containingPath}) {
-    _parseDescription(description);
-    return PackageRef(name, this, description);
+  PackageRef parseRef(String name, description,
+      {String containingPath, LanguageVersion languageVersion}) {
+    return PackageRef(
+        name, this, _parseDescription(name, description, languageVersion));
   }
 
   @override
   PackageId parseId(String name, Version version, description,
       {String containingPath}) {
-    _parseDescription(description);
-    return PackageId(name, this, version, description);
+    // Old pub versions only wrote `description: <pkg>` into the lock file.
+    if (description is String) {
+      assert(description == name);
+      return PackageId(
+          name, this, version, _HostedDescription(name, defaultUrl));
+    }
+
+    final serializedDescription = (description as Map).cast<String, String>();
+
+    return PackageId(
+      name,
+      this,
+      version,
+      _HostedDescription(serializedDescription['name'],
+          Uri.parse(serializedDescription['url'])),
+    );
   }
 
+  _HostedDescription _asDescription(desc) => desc as _HostedDescription;
+
   /// Parses the description for a package.
   ///
   /// If the package parses correctly, this returns a (name, url) pair. If not,
   /// this throws a descriptive FormatException.
-  Pair<String, Uri> _parseDescription(description) {
+  _HostedDescription _parseDescription(
+      String packageName, description, LanguageVersion version) {
```

Comment:
nit: can we rename this local - it can be confused with the version of the package.
```suggestion
      String packageName, description, LanguageVersion languageVersion) {
```